Feature/1.2.2.5/db connection from frontend

sharrief opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This refactor cleans up a lot of code in preparation for streamlining the e2e input demo. 
As a lot of setup/config related changes have been made, we are looking to merge this PR so that other branches have access to the critical pieces, like the updated types and ClientService.ts

- Removes unused code in AnimalService/Dialog
- Updates types.ts so that all entity types are based on the generated DB types
- Fixes supabase queries in ClientServices.ts 
- Moves other classes from ClientService.ts into their own files in src/services/lib
- Fixes tests for ClientService.ts
- Adds dev container configuration files to enable local development inside a container
- Updates the README with instructions on getting repository running locally, incl. Google OAuth setup
- Adds stub code for Supabase Edge functions in Deno. See supabase/functions 

**Running tests locally require that a local supabase instance has been setup and is running, and that the database.types.ts matches the local instance schema.** 

To get the updated types
`npm run supa-login` if you've never run it before
`npm run supa-start` to start the supabase db and services
`npm run gen-types` to generate the types from the db

## Related Tickets & Documents

[1.2.2.0 Wire up input form for saving to the database](https://projects.digitalaidseattle.org/project-tasks-details?recordId=recm4ng99AfweJQak)

## QA Instructions, Screenshots, Recordings

Instructions for local setup are in the updated README. 

### UI accessibility checklist
_If your PR includes UI changes, please utilize this checklist:_
- [ ] Semantic HTML implemented?
- [ ] Keyboard operability supported?
- [ ] Color contrast tested?

## Added/updated tests?
_We are aiming for 100% statement coverage._

- [ ] Yes
- [x] No, and this is why: This code is only refactoring existing code. More coverage will be added in future PRs
- [ ] I need help with writing tests

## [optional] Are there any post-deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?
### When we realize that several of us are working on the same file
![When we realize that several of us are working on the same file](https://thecodinglove.com/content/050/CSMAz3N.gif)

samhhenderson commented:
I modified a couple small things (see the commit) in order to test the full flow. Before deploying I wanted to check if you know why the Vercel deployment is failing (I don't have access when I try to look at Details).

samhhenderson reviewed (changes requested):
Let's try to figure out why the Vercel build is failing

sharrief reviewed (commented):
> Let's try to figure out why the Vercel build is failing 

Its because the production DB is out of sync with my local DB. The issue is  service_categories, the insert schema in prod has changed_at as a requried field. But my changes have it and all the other fields as optional. I think it should be optional since the DB will put in the default value of now() anyways. But I don't know how to change the database.types.ts file that the build is using.

jnakaso reviewed (changes requested):
Lot going on here, more than just db connection from frontend.  Can we break up this PR?  One for breaking up types into individual files for instance.

sharrief commented:
> Lot going on here, more than just db connection from frontend.  Can we break up this PR?  One for breaking up types into individual files for instance.

Really this PR is just for updating the types. Didn't make any progress on front end. But the types are using in a lot of files so lots of files were touched. Also I fixed places where the code was broken because the types weren't used correctly

jnakaso commented:
> > Lot going on here, more than just db connection from frontend.  Can we break up this PR?  One for breaking up types into individual files for instance.
> 
> Really this PR is just for updating the types. Didn't make any progress on front end. But the types are using in a lot of files so lots of files were touched. Also I fixed places where the code was broken because the types weren't used correctly

There are updates to package, readme, migration, tests utils, type name changes, edge-function support...those could be broken up.

jnakaso commented on utils/testSetup.ts:
Test files should probably be move to test direction.

sharrief replied:
Will ping in Slack on this topic.

sharrief replied:
Also forgot that specifically these are just test setup files, not tests themselves. When you put these files into the __tests__ folder then Jest complains and says there are no tests present in these files. Eventually we will mock the supabase client and code like this will go into __mocks__, so that's the long term plan I think

sharrief commented:
> > > Lot going on here, more than just db connection from frontend.  Can we break up this PR?  One for breaking up types into individual files for instance.
> > 
> > 
> > Really this PR is just for updating the types. Didn't make any progress on front end. But the types are using in a lot of files so lots of files were touched. Also I fixed places where the code was broken because the types weren't used correctly
> 
> There are updates to package, readme, migration, tests utils, type name changes, edge-function support...those could be broken up.

Its a large change I admit, and I'll do my best to explain why. For prerequisite context: everyone's work relies on having the correct types for the schema, otherwise we end up designing code that make incorrect assumptions about the data and parameters. 

So the goal of this PR is primarily to get the correct types introduced. As a part of doing that, I found that lots of the code was broken. And in order to verify that my changes work, I had to fix these broken parts of the code.

1. In order to generate the types I needed to get supabase running locally. Our migration files were out of date in this branch, so I updated those, as having the right schema is a pre-req for generating the correct types. 

2. I also updated the readme with instructions that help everyone else get their environment running with the same parts. This seemed a necessary step to make sure these changes can be validated by someone else and so everyone can do further development. I added some scripts to package.json so others can easily start supabase locally and generate the types.

3. After generating the types (database.types.ts) I then created aliases for those types and tried to chose meaningful names. I think we can still come up with better names. Nouns like "client", "request" and "service" have many interpretations in software, so I think we should move away from them. Perhaps "users" and "ticket" would be better. For this PR I simply decided to make it more clear these types were referring to the schema and not to a class or a some other generic built-in type.

4. After getting the correct types in place, TypeScript showed that many areas of the code were making incorrect assumptions about the data  types, so I fixed those.

5. ClientService.ts was crowded, it had multiple classes, enums and functions. I think some of it can be deleted after updating other code to use the appropriate types from the DB. Ex: NewClientRequest can probably be replaced and I did delete AnimalService/Dialog since it wasn't used anywhere. But for now I mostly put those into separate files in src/lib

6. I can remove the Deno changes, and I'll ping in Slack about findings on this approach after some initial investigation.

In order to break some of these related items up into separate PRs, I would have to create a branch for each. Alternatively, I could hold some changes on my machine and push them incrementally in a series of PRs. But both those paths would hinder my and other's ability to build on top of this work. I truly understand that this is a large PR, and that in general large PRs should be avoided. However, this PR appears large by nature of catching up on some of our technical debt.

sharrief commented:
Closing to re-open several smaller PRs
